Overhaul Object's Documentation

Mickeon opened the pull request:
### This PR is touchy

Closes https://github.com/godotengine/godot-docs/issues/4894
Closes https://github.com/godotengine/godot-docs/issues/5139
Closes https://github.com/godotengine/godot-docs/issues/3827 (???)
... and maybe some more?

Continuation of #67072, #67100, #67208, #67718...

Here a big list of changes which, unlike the previous PRs, is a lot more detailed. Although, I have to admit, I got tired of writing it due to repetition. 
Mainly, it's meant to represent my line of reasoning behind every change. Perhaps, someone reading this may take it as a future point of reference. Not just that, but some information deemed important by the maintainers could have been lost during the rewrite. Criticism is encouraged.

### General
- Made use of _[param]_ and several other strong references more often.
- Corrected typos.
- Stripped awkward, needlessly long sentences.
- Simplified wording where it felt necessary. For example:
   - _"Contrarily to"_ -> _"Unlike"_;
   - _"taken into account"_ -> "included"_.
-  Changed wording where technical terms (also used in the rest of the documentation) are more appropriate:
   - _"can take"_ -> _"supports"_;
- Used synonyms when words could be mistaken for another concept of Godot's API. For example:
   - _"parent classes"_ -> _"inherited classes"_;
   - _"plural object"_ -> _"message's subject"_
- Added way more examples.
- The spacing of comments in the examples has been adjusted to improve readability.
   - This is not a hard rule defined in the **ClassRef** guidelines. Although, for reasons stated above, it could totally be. However, see https://github.com/godotengine/godot-docs/issues/6328#issuecomment-1290696182.
- Tried to refer to "signal" as "signal name" sometimes.
    - This is because **Signal** can be a Variant type in Godot 4, but the API may still need to refer to them by name.

### Methods
#### 
- The leading description has been completely overhauled.
    - _"a built-in type" 
    What? Node type? Class type?_ Unclear. It's **Variant** type.
    - _"Objects do not manage memory."_, 
    Then, **literally** the next paragraph began with _"Some classes that extend Object add memory management"_.
    - _"Objects export properties, which are mainly useful for storage and editing, but not really so much in programming."_ 
  **W-what!?**
    - The `in` operator checks for properties, _methods_, and _signals_ _(try this I dare you)_.
    - Introduced basic concept of inheritance.
    - Introduced Scripts.
    - Introduced Notifications.
    - Introduced Metadata.
- Reworked **`_get_property_list`**.
    - Added important notes.
- Reworded **`_init`**.
    - Specified to be the script, not the Object.
- Reworded **`can_translate_messages`**.
    - "if the object can translate strings".
    Eh? What strings?
- Reworded **`connect`**.
    - Detail returned **Error** constant.
    - It's better to reword it for now....
    The description of this method is... so needlessly long, verbose, and incredibly horrifying to look at! Despite this, ironically, this isn't even the recommended way to connect signals anymore. This may need to be seen in another PR, it's just... ridiculous. It even explains... `emit_signal`? Why not putting the explanation there, then!? Aaah...
    - _"One can pass a binds Array to Object.connect()"_ this is just plainly false in Godot 4.
- Reworked **`emit_signal`**.
    - Detail returned error Error constant.
- Reworked **`get_incoming_connections`**.
    - It was entirely wrong and not updated from **3.x**.
- Reworked **`get_indexed`** and **`set_indexed`**.
    - Add missing note for C# about Built-in pascal_case properties.
- Reworked **`get_method_list`**.
   - Detail the returned  **Dictionary** **Array**.
     Closes https://github.com/godotengine/godot-docs/issues/4894.
- Reworked **`get_property_list`**.
    - Added more detail to the returned **Dictionary** **Array**.
- Reworked **`get_signal_list`**.
    - Added more detail to the returned **Dictionary** **Array** 
    (Which is incredibly strange by the way).
- Reworded **`notify_property_list_changed`**.
    - _"Notify the editor that the property list has changed by emitting the property_list_changed signal"_.
    These lines have been swapped, because all this method does is emitting the signal.
- Reworded **`set_script`**.
    - _"Assign script"_ -> _"Attach script"_. 
    Although both terms are correct, _"attach"_ matches the more familiar verb used in the Scene Tree Editor.
    - _"Each object can have a single script assigned to it"_.
    Very much sounded like a leftover from when Godot used to support multiple scripts.
- Reworded **`to_string`**.
    - _"defaults to "ClassName:RID""_
    This was outdated.

### Signals
- Add missing description to **`property_list_changed`**.
- Recommend deferring **`script_changed`** on some cases.
    - This is because when the signal is emitted, the new script is not yet initialized.

### Constants
- Specified that **NOTIFICATION_POSTINITIALIZE** is pretty much used internally.
    - You can't receive it from **Script**, because normally it is sent even before a **Script**'s `_init()` is even called...
- Made all **CONNECT** constant descriptions consistent.
    - They all begin like _"[adjective] connections do this thing"_. I am not sure they are the best way to word them, but it is consistent.
    
### TODO:
- [x] Finish rewriting leading description.
- [x] ~Possibly adjust `connect` to reduce... overwhelming-ness.~ Barely... May not be doable for now.
- [x] ~Document `_set` and `_get`'s _very interesting_ use cases (when it is actually called)~. Can't be done as I wished. It almost feels like the behavior is bugged.
- [x] Refine `get_property_list` and `_get_property_list`.

I recommend taking a look at the documentation by building from this PR, to read the documentation from the Editor itself. Feedback is welcome.

Mickeon committed:
Overhaul Object's Documentation

Iniquitatis commented on doc/classes/Object.xml:
```diff
@@ -4,30 +4,41 @@
 		Objects export properties, which are mainly useful for storage and editing, but not really so much in programming. Properties are exported in [method _get_property_list] and handled in [method _get] and [method _set]. However, scripting languages and C++ have simpler means to export them.
-		Property membership can be tested directly in GDScript using [code]in[/code]:
+
+		Objects can extend their functionality by attaching a [Script]. Likewise, all, most of the documentation assumes
```
Unfinished text in lines 13, 15, 19.

Iniquitatis commented on doc/classes/Object.xml:
```diff
@@ -260,9 +290,11 @@
 				For legacy or language-specific reasons, there are also several ways to bind parameters to signals. One can pass a [code]binds[/code] [Array] to [method Object.connect] or [method Signal.connect], or use the recommended [method Callable.bind] method to create a new callable from an existing one, with the given parameter binds.
 				One can also pass additional parameters when emitting the signal with [method emit_signal]. The examples below show the relationship between those two types of parameters.
+				
+				While there are multiple ways to bind parameters to connections, the recommended syntax
```
Also unfinished.

Mickeon replied:
Did you notice this PR is still a draft?

Iniquitatis replied:
Sure. If this is intended for now, then sorry and feel free to ignore the comments.

_These lines just seemed as a consequence of some kind of text editor bug (?), or whatever._

Mickeon replied:
> These lines just seemed as a consequence of some kind of text editor bug (?), or whatever.

In reality they're dummy lines, partial attempts to rewrite the same thing in a different way to see what is best. I have not removed and tidied up a few just yet.

timothyqiu commented on doc/classes/Object.xml:
```diff
@@ -510,88 +610,118 @@
 			<description>
-				Assigns a script to the object. Each object can have a single script assigned to it, which are used to extend its functionality.
-				If the object already had a script, the previous script instance will be freed and its variables and state will be lost. The new script's [method _init] method will be called.
+				Attaches [param script] to the object, and calls the script's [method _init] to instantiate it. A [Script] is used to extend the object's functionality.
+				If a script already exists, its instance is detached and its property values and state are lost, unless it's the same as [param script]. Built-in property values are still kept.
```
Same problem as `_init()` description.

This method does not call the script's `_init()` to instantiate it. It instantiates the script and attaches the script instance. The `_init()` call is the result of instantiating the script.

> unless it's the same as [param script]

Skipping for unchanged value is a common behavior of setters I think.

Mickeon replied:
> Skipping for unchanged value is a common behavior of setters I think.

It is, indeed, but I thought that the effects of changing a script were more... "catastrophic" than most setters, so it needed a mention. It can (?) be removed since some user that needs this information can probably notice it with its own eyes...??

timothyqiu commented on doc/classes/Object.xml:
```diff
@@ -601,43 +731,45 @@
 	<constants>
 		<constant name="NOTIFICATION_POSTINITIALIZE" value="0">
-			Called right when the object is initialized. Not available in script.
+			Notification received when the object is initialized, before its script's [method _init] is called. Used internally.
```
```suggestion
			Notification received when the object is initialized, before its script is attached. Used internally.
```

timothyqiu commented on doc/classes/Object.xml:
```diff
@@ -601,43 +731,45 @@
 		</constant>
 		<constant name="CONNECT_PERSIST" value="2" enum="ConnectFlags">
-			Persisting connections are saved when the object is serialized to file.
+			Persisting connections are stored when the object is serialised (such as when using [method PackedScene.pack]). In the editor, connections created through the Node dock are always persisting.
```
```suggestion
			Persisting connections are stored when the object is serialized (such as when using [method PackedScene.pack]). In the editor, connections created through the Node dock are always persisting.
```

Since all other places spell this way.

timothyqiu commented on doc/classes/Object.xml:
```diff
@@ -601,43 +731,45 @@
 		</constant>
 		<constant name="CONNECT_REFERENCE_COUNTED" value="8" enum="ConnectFlags">
-			Connect a signal as reference-counted. This means that a given signal can be connected several times to the same target, and will only be fully disconnected once no references are left.
+			Reference-counted connections can be assigned to the same [Callable] multiple times. Calling the [method disconnect] method subtracts one connection at the time, until no connections remain.
```
English is not my native language. I think "until no connections remain" is redundant and sounds wrong to me.

timothyqiu replied:
To me, `"will only be fully disconnected once no references are left"` makes sense because it means something like this:

```
if current_reference_count <= 0:
    you_are_fully_disconnected()
```

Although kind of silly, I think `"Calling the method subtracts one connection at the time, until no connections remain"` can be interpreted as:

```
func method():
    while still_have_connections():
        connection -= 1
```

Mickeon replied:
I agree with @timothyqiu . It may sound like "babyfying" the explanation, but CONNECT_REFERENCE_COUNTED is a pretty simple concept with a deserved simple explanation, that can be easily tested to be fully understood.

Mickeon replied:
Not sure about _preferring_ the previous explanation, but I agree on the principle.

Mickeon replied:
Accidentally pushed early, but the description of this one has been changed. You may review again.

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -53,62 +64,80 @@
 			<description>
-				Virtual method which can be overridden to customize the return value of [method set].
-				Sets a property. Returns [code]true[/code] if the [param property] exists.
+				Virtual method which can be overridden to customize the behavior of [method set] for the given [param property].
```
I was waiting to adjust it to do some testing myself. Howeve, when I got to it, at least in my current build, it really seemed like `_set` and `_get` did not exist. No matter what I did (and unlike 3.x) I was unable to trigger them.

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -53,62 +64,80 @@
 			<description>
-				Called when the object is initialized in memory. Can be defined to take in parameters, that are passed in when constructing.
-				[b]Note:[/b] If [method _init] is defined with required parameters, then explicit construction is the only valid means of creating an Object of the class. If any other means (such as [method PackedScene.instantiate]) is used, then initialization will fail.
+				Called when the object's script is initialized in memory. Can be defined to take in parameters. This method is similar to a constructor in most programming languages.
```
I was reluctant to introduce the concept of script instances, with no explanation in the leading description. From the user's perspective, there's little need to know about it, because it's easier to think of the Script as a Resource that can be used to extend Object over Script Vs. Script Instance. It doesn't mean I was going to strip away every mention of it, but it requires some thought.

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -170,7 +200,7 @@
+				Connects a [param signal] to a [param callable]. Optional [param flags] can be also added to configure the connection's behavior (see [enum ConnectFlags] constants).
 				A signal can only be connected once to a [Callable]. It will print an error if already connected, unless the signal was connected with [constant CONNECT_REFERENCE_COUNTED]. To avoid this, first, use [method is_connected] to check for existing connections.
 				If the callable's target is destroyed in the game's lifecycle, the connection will be lost.
 				[b]Examples with recommended syntax:[/b]
```
Some of it shouldn't be outright moved elsewhere because other languages could need the example as a point of reference.

Honestly, the entire block is absurd. There's little to no point in showing the many different ways you can connect a signal, while only one is recommend. It's also easily the biggest description in the Class Reference. For its importance, that may be very understandable, but it's so large it may need its own PR.

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -4,30 +4,41 @@
 	<description>
-		Every class which is not a built-in type inherits from this class.
-		You can construct Objects from scripting languages, using [code]Object.new()[/code] in GDScript, or [code]new Object[/code] in C#.
+		Every class which is not a built-in [Variant] type inherits from this class.
```
There's no consistent definition for these types in the documentation. Sometimes they're "structures", mostly "types" or "Variant types" at most. The one major exception to it is Object for being... special...

Mickeon replied:
Updated the leading description. It is still unfinished and requires one more refinement step.

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -334,146 +366,204 @@
-				Returns the object's metadata entry for the given [param name].
-				Throws error if the entry does not exist, unless [param default] is not [code]null[/code] (in which case the default value will be returned). See also [method has_meta], [method set_meta] and [method remove_meta].
-				[b]Note:[/b] Metadata that has a [param name] starting with an underscore ([code]_[/code]) is considered editor-only. Editor-only metadata is not displayed in the inspector and should not be edited.
+				Returns the object's metadata value for the given entry [param name]. If the entry does not exist and [param default] is omitted, an error is generated. Otherwise, returns [param default].
```
It is not technically accurate because it returns whatever **default** is on error, but... `null == null`. It's probably worth mentioning the null regardless...

Mickeon commented:
Still unfinished. Introduced metadata in the leading description. it's a core concept, sounds well deserved. Refined `_get_property_list()` with a really nice example, I find.

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -4,30 +4,42 @@
-		[b]Note:[/b] Unlike references to a [RefCounted], references to an Object stored in a variable can become invalid without warning. Therefore, it's recommended to use [RefCounted] for data classes instead of [Object].
-		[b]Note:[/b] The [code]script[/code] property is not exposed like most properties, but it does have a setter and getter ([code]set_script()[/code] and [code]get_script()[/code]).
+		Objects also receive notifications, frequently sent by the engine for common events. See [method _notification].
+		Every object may also contain metadata (data within data), with [method set_meta]. It can be useful to store information that the object itself does not depend on. To keep the code clean, making excessive use of metadata is discouraged.
```
Not technically incorrect though!

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -4,30 +4,41 @@
+		Customizing properties this may allows them
+		Objects may contain properties. The attributes may be customised with [method _get_property_list]. This allows any property to be exposed to the editor, saved to disk,
+
+		In GDScript, you can check if a given property, method, or signal exists in an object with [code]in[/code]:
```
Updated the leading description. The part potentially mentioning this is not finished. It's very difficult to encourage duck typing in the description...

Mickeon commented:
The Description is still WIP and requires some rearranging, but it's good to read to get an idea of the thought process behind it.

Mickeon commented:
The leading description is in a pretty good state now. Ready for review.

anvilfolk commented on doc/classes/Object.xml:
```diff
@@ -1,33 +1,38 @@
+
+		Objects can have a [Script] attached to them. Once the [Script] is instantiated, it effectively acts as an extension to the base class, allowing it to define and inherit new properties, methods and signals.
+		Inside a [Script], virtual methods can be overriden to change behavior. In particular, [method _get_property_list] may be overriden to customize properties in several ways. This allows them to be available to the editor, display as lists of options, sub-divide into groups, save on disk, etc. Scripting languages offer easier ways to customize properties, such as with the [annotation @GDScript.@e ...
+		Because Godot is very dynamic, there can be occasions where, for example, a property may not exist. To prevent run-time errors, see methods such as [method set], [method get], [method call], [method has_method], [method has_signal] etc. Note that these methods are much slower than a direct reference to the members. In GDScript, you can also check if a given property, method, or signal name exis ...
```
It might be more helpful to say what makes Godot/Object dynamic, rather than that it simply is dynamic. Something like Object properties and methods being able to change at runtime to fit app/game needs, and so it is useful to have ways to check and set those properties, using the the methods you're already listing!

anvilfolk replied:
Definitely better, though for newbies (like meee) explicitly saying that an Object's properties can change at runtime is helpful in that it just settles the issue for the reader. That way it doesn't have to be inferred from the fact that 1) the script defines the properties and the script can change, and 2) the _set and _get methods actually exist!

Mickeon replied:
I am struggling to understand this is a positive comment or not, could you elaborate?

anvilfolk replied:
I think what I'm saying is that I'd personally find useful to see this change:

`Godot is very dynamic. An object's script can be changed at run-time, and all objects can be passed into any method with Object parameters.`

to

`Godot is very dynamic. An object's script, and therefore its properties and methods, can be changed at run-time.`

This says the properties and methods change at runtime explicitly, which I think is helpful to assert early and clearly. I think the part about objects isn't necessary, since it's not specific to Godot, just programming languages.

Hopefully that clarified it!

Mickeon replied:
It did very much. I modified the description once more with something similar to your sentence.

anvilfolk replied:
Looks wonderful! <3

anvilfolk commented on doc/classes/Object.xml:
```diff
@@ -1,33 +1,38 @@
-		Objects also receive notifications. Notifications are a simple way to notify the object about different events, so they can all be handled together. See [method _notification].
-		[b]Note:[/b] Unlike references to a [RefCounted], references to an Object stored in a variable can become invalid without warning. Therefore, it's recommended to use [RefCounted] for data classes instead of [Object].
-		[b]Note:[/b] The [code]script[/code] property is not exposed like most properties, but it does have a setter and getter ([code]set_script()[/code] and [code]get_script()[/code]).
+		Objects also receive notifications, frequently sent by the engine for common events. See [method _notification].
```
An example of event could be great, a window resized, or a physics collision?

anvilfolk replied:
Perfect :)

anvilfolk commented on doc/classes/Object.xml:
```diff
@@ -1,33 +1,38 @@
+		Objects also receive notifications, frequently sent by the engine for common events. See [method _notification].
+		Every object may also contain metadata (data about data), with [method set_meta]. It can be useful to store information that the object itself does not depend on. To keep your code clean, making excessive use of metadata is discouraged.
+		[b]Note:[/b] Unlike references to a [RefCounted], references to an Object stored in a variable can become invalid unexpectedly. To check for this, use [method @GlobalScope.is_instance_valid]. It's recommended to inherit from [RefCounted] for classes storing data instead of [Object].
+		[b]Note:[/b] The [code]script[/code] property is not exposed normally. To set or get an object's [Script] in code, use [method set_script] and [method get_script], respectively.
```
Not exposed normally reads like it isn't often that it is exposed - kind of like that it is unusual that it is exposed. I think I'd prefer the "not exposed like other properties".

anvilfolk commented on doc/classes/Object.xml:
```diff
@@ -38,77 +43,119 @@
-				Virtual method which can be overridden to customize the return value of [method get_property_list].
-				Returns the object's property list as an [Array] of dictionaries.
-				Each property's [Dictionary] must contain at least [code]name: String[/code] and [code]type: int[/code] (see [enum Variant.Type]) entries. Optionally, it can also include [code]hint: int[/code] (see [enum PropertyHint]), [code]hint_string: String[/code], and [code]usage: int[/code] (see [enum PropertyUsageFlags]).
+				Override this method to customize how script properties should be handled by the engine.
+				Should return a property list, as an [Array] of dictionaries. The result is added to the array of [method get_property_list], and should be formatted in the same way. Each [Dictionary] must at least contain the [code]name[/code] and [code]type[/code] entries.
+				The example below displays [code]hammer_type[/code] in the Inspector dock, and saves it, only if [code]using_hammer[/code] is [code]true[/code]:
+				[codeblock]
+				@tool
+				extends Node2D
+
+				@export var using_hammer = false:
+				    set(value):
+				        using_hammer = value
+				        notify_property_list_changed()
+				var hammer_type = 0
+
+				func _get_property_list():
+				    var properties = []
+
+				    if can_jump:
+				        properties.append({
+				            "name": "hammer_type",
+				            "type": TYPE_INT,
+				            "usage": PROPERTY_USAGE_DEFAULT,
+				            "hint": PROPERTY_HINT_ENUM,
+				            "hint_string": "Wooden,Iron,Golden,Enchanted"
+				        })
+
+				    return properties
+				[/codeblock]
+				[b]Note:[/b] This method is intended for advanced purposes. For most common use cases, the scripting languages offer easier ways to handle properties. See [annotation @GDScript.@export], [annotation @GDScript.@export_enum], [annotation @GDScript.@export_group], etc.
+				[b]Note:[/b] If the object's script is not [annotation @GDScript.@tool], this method will not be called in the editor.
```
Love that there's an example showing how this is used!!! Would've been super helpful to me just a few days ago :)

I think line 68 should be `if using_hammer:` rather than `if can_jump:`!

I think it's still useful to have links to Variant.Type, PropertyHint, etc like in the original.

Mickeon replied:
I decided not to mention them because all of the information about what this it is present in `get_property_list()`.
I personally believe that all descriptions should be as "self-sustaining" as possible, with very little redirecting around, but this one is an exception because... There's a lot of words.
It's not outstandingly bad to add them back, but it may make the method look more complicated than it *already* is.

anvilfolk replied:
That's fair, and if people are interested in the enums that do show up like PROPERTY_HINT_ENUM, they can also manually search for them in the documentation anyway!

timothyqiu commented on doc/classes/Object.xml:
```diff
@@ -1,33 +1,38 @@
-		Property membership can be tested directly in GDScript using [code]in[/code]:
+		An advanced [Variant] type. All classes in the engine inherit from Object. Each class may define new properties, methods or signals, which are available to all inheriting classes. For example, a [Sprite2D] instance is able to call [method Node.add_child] because it inherits from [Node].
+		You can create new instances, using [code]Object.new()[/code] in GDScript, or [code]new Object[/code] in C#.
+		To delete an Object instance, call [method free]. This prevents memory leaks, and is necessary for most classes inheriting Object, because they do not manage memory on their own. [RefCounted] is the biggest exception to this. [Node], another essential object, also deletes its children when freed.
```
> To delete an Object instance, call [method free]. This prevents memory leaks...

Nitpick: the memory leak part is a bit strange. It sounds like:

```gdscript
if you_want_delete_an_object():
    please_call_free()
    assert(not have_memory_leaks())
else:
    pass  # Most cases of memory leaks happen here, i.e. you forget to delete it.
```

This is technically correct, but memory leaks are caused by not deleting objects, not a result of not calling `free()`.

```suggestion
		To delete an Object instance manually, call [method free] or [method queue_free]. This is necessary for most classes inheriting Object, because they do not manage memory on their own. There are a few classes that perform memory management. For example, [RefCounted] deletes itself when no longer referenced and [Node] deletes its children when freed. Please make sure an object is deleted when it is no longer needed, as this will prevent memory leaks.
```

Mickeon replied:
I know it may sound incredibly strange but Object by itself does not have `queue_free`.

timothyqiu replied:
Ah, yeah. I forget it's a Node method.

timothyqiu commented on doc/classes/Object.xml:
```diff
@@ -1,33 +1,38 @@
+		To delete an Object instance, call [method free]. This prevents memory leaks, and is necessary for most classes inheriting Object, because they do not manage memory on their own. [RefCounted] is the biggest exception to this. [Node], another essential object, also deletes its children when freed.
+
+		Objects can have a [Script] attached to them. Once the [Script] is instantiated, it effectively acts as an extension to the base class, allowing it to define and inherit new properties, methods and signals.
+		Inside a [Script], virtual methods can be overriden to change behavior. In particular, [method _get_property_list] may be overriden to customize properties in several ways. This allows them to be available to the editor, display as lists of options, sub-divide into groups, save on disk, etc. Scripting languages offer easier ways to customize properties, such as with the [annotation @GDScript.@e ...
```
I think this paragraph is intended to describe how virtual methods work. But too much is spent on describing `_get_property_list` instead.

Mickeon replied:
The opposite is true

Mickeon commented:
Another one bites the dust.

anvilfolk commented on doc/classes/Object.xml:
```diff
@@ -38,77 +42,119 @@
 			<param index="0" name="method" type="StringName" />
 			<description>
-				Calls the [param method] on the object and returns the result. This method supports a variable number of arguments, so parameters are passed as a comma separated list. Example:
+				Calls the [param method] on the object and returns the result. This method supports a variable number of arguments, so parameters can be passed as a comma separated list.
```
Any chance we could get an example here? The examples added elsewhere were really helpful, and this one feels subtle enough to warrant one :)

anvilfolk commented on doc/classes/Object.xml:
```diff
@@ -170,23 +217,23 @@
 				    button.button_down.connect(_on_button_down)
 
-				    # This assumes that a `Player` class exists which defines a `hit` signal.
+				    # This assumes that a `Player` class which defines a `hit` signal exists.
```
The same sentence exists below on line 252, could use the same edit for consistency!

anvilfolk replied:
Same on 316!

Mickeon replied:
I kind of reverted the wording a bit here and added a comma. In retrospective it didn't flow right.

anvilfolk commented on doc/classes/Object.xml:
```diff
@@ -276,8 +323,8 @@
-				# Four arguments, since we pass two when emitting (hit_by, level)
-				# and two when connecting (weapon_type, damage).
+				# We pass two arguments when emitting (`hit_by`, `level`),
+				# and two more arguments when connecting (`weapon_type`, `damage`).
```
Whatever the decision, it also applies to comments on line 346.

Mickeon replied:
It's actually the result of `Callable.bind()`, yes. It seems like just adding the verb makes sense here.

Mickeon commented:
Rebased after https://github.com/godotengine/godot/pull/68120.

mhilbrunner reviewed (approved):
Can't believe I've spotted nothing so far, but so I did, and I read it twice now. Great changes and fixes in here!

KoBeWi commented on doc/classes/Object.xml:
```diff
@@ -38,78 +42,119 @@
+				Override this method to customize how script properties should be handled by the engine.
+				Should return a property list, as an [Array] of dictionaries. The result is added to the array of [method get_property_list], and should be formatted in the same way. Each [Dictionary] must at least contain the [code]name[/code] and [code]type[/code] entries.
+				The example below displays [code]hammer_type[/code] in the Inspector dock, and serializes it, only if [code]using_hammer[/code] is [code]true[/code]:
+				[codeblock]
+				@tool
+				extends Node2D
+
+				@export var using_hammer = false:
+				    set(value):
+				        using_hammer = value
+				        notify_property_list_changed()
+				var hammer_type = 0
+
+				func _get_property_list():
+				    var properties = []
+
+				    if using_hammer:
+				        properties.append({
+				            "name": "hammer_type",
+				            "type": TYPE_INT,
+				            "usage": PROPERTY_USAGE_DEFAULT,
+				            "hint": PROPERTY_HINT_ENUM,
+				            "hint_string": "Wooden,Iron,Golden,Enchanted"
+				        })
+
+				    return properties
+				[/codeblock]
```
This example is not complete, because it's missing `_set()` and `_get()`. The property will just appear in the inspector without being saved. See https://github.com/godotengine/godot/issues/68797#issuecomment-1321258302

Mickeon replied:
Feels like an oversight, I don't recall this happening in 3.x...

KoBeWi replied:
It was always like that. The properties you return in `_get_property_list()` don't even have to exist.

Mickeon replied:
They do not have to exist, but it was under my assumption that it would save.

In fact, I just tested the snippet of code, and the `hammer_type` property does save completely fine, at least within the currently edited scene.

KoBeWi replied:
It seems to be always 0 though.
I think the method should at least mention `_get()` and `_set()`, because they are closely related. (and these 2 methods should have some examples)

Mickeon replied:
I was trying to provide examples for `_get()` and `_set()`, and they would not work as I expected them to (nor as how they worked in 3.x), and I can't tell if it's me missing something or an oversight.

KoBeWi replied:
In the issue I linked there is a working example.

YuriSizov replied:
As mentioned in the linked issue, you don't need `_get` and `_set` if your `_get_property_list` defines a property with the same name as an existing class member. You only need those two methods for purely computed properties.

See https://github.com/Arnklit/ShellFurGodot/blob/main/addons/shell_fur/shell_fur_manager.gd as an example of this being used in a 3.x plugin from a couple of years ago.

So this code snippet should be valid in principle, though I've not tested it to see if there are some other issues.

Mickeon commented:
If we want to consider my comment as a patch note, I modified the example of `_get_property_list()` as [it may or may not have been problematic](https://chat.godotengine.org/channel/core?msg=xY8GmtL9yvzwjYeWM), and added examples and further explanation for `_get()` and `_set()`

Mickeon commented:
If we want to consider my comment as a patch note, I modified the example of `_get_property_list()` as [it may or may not have been problematic](https://chat.godotengine.org/channel/core?msg=xY8GmtL9yvzwjYeWM), and added examples and further explanation for `_get()` and `_set()`

YuriSizov commented on doc/classes/Object.xml:
```diff
@@ -1,33 +1,37 @@
 <class name="Object" version="4.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
 	<brief_description>
-		Base class for all non-built-in types.
+		Base class for all objects in the engine.
```
As I mentioned on RC, I'm not sure it's a good idea to use the word that you are trying to describe in its own description. Though the "Object" class in Godot and the concept of "object" in OOP are not the exact same thing, the reader may not be aware of that. Changing it for a better term might prove hard due to how basic it is. I'd suggest a slight adjustment at least:

```suggestion
		Base class for all object types in the engine.
```

YuriSizov replied:
Or maybe just use "classes" since this is what we do in the full description.

```suggestion
		Base class for all other classes in the engine.
```

YuriSizov commented on doc/classes/Object.xml:
```diff
@@ -1,33 +1,37 @@
+		To delete an Object instance, call [method free]. This is necessary for most classes inheriting Object, because they do not manage memory on their own, and will otherwise cause memory leaks when no longer in use. There are a few classes that perform memory management. For example, [RefCounted] (and by extension [Resource]) deletes itself when no longer referenced, and [Node] deletes its childre ...
+		Objects can have a [Script] attached to them. Once the [Script] is instantiated, it effectively acts as an extension to the base class, allowing it to define and inherit new properties, methods and signals.
+		Inside a [Script], [method _get_property_list] may be overriden to customize properties in several ways. This allows them to be available to the editor, display as lists of options, sub-divide into groups, save on disk, etc. Scripting languages offer easier ways to customize properties, such as with the [annotation @GDScript.@export] annotation.
+		Godot is very dynamic. An object's script, and therefore its properties, methods and signals, can be changed at run-time. Because of this, there can be occasions where, for example, a property required by a method may not exist. To prevent run-time errors, see methods such as [method set], [method get], [method call], [method has_method], [method has_signal] etc. Note that these methods are [b] ...
```
```suggestion
		Godot is very dynamic. An object's script, and therefore its properties, methods and signals, can be changed at run-time. Because of this, there can be occasions where, for example, a property required by a method may not exist. To prevent run-time errors, see methods such as [method set], [method get], [method call], [method has_method], [method has_signal], etc. Note that these methods are [b]much[/b] slower than direct references. In GDScript, you can also check if a given property, method, or signal name exists in an object with the [code]in[/code] operator:
```

Missing comma. Also the parts starting with "In GDScript" should probably be its own paragraph. Also also, it's weird to say "In GDScript" and give examples in both GDScript and C#, with the latter using a different syntax 🙂

YuriSizov commented on doc/classes/Object.xml:
```diff
@@ -1,33 +1,37 @@
-		[b]Note:[/b] The [code]script[/code] property is not exposed like most properties, but it does have a setter and getter ([code]set_script()[/code] and [code]get_script()[/code]).
+		Notifications are [int] constants commonly sent and received by objects. For example, on every rendered frame, the [SceneTree] notifies nodes inside the tree with a [constant Node.NOTIFICATION_PROCESS]. The nodes receive it and may call [method Node._process] to update. To make use of notifications, see [method notification] and [method _notification].
+		Lastly, every object can also contain metadata (data about data). [method set_meta] can be useful to store information that the object itself does not depend on. To keep your code clean, making excessive use of metadata is discouraged.
+		[b]Note:[/b] Unlike references to a [RefCounted], references to an Object stored in a variable can become invalid unexpectedly. To check for this, use [method @GlobalScope.is_instance_valid]. It's recommended to inherit from [RefCounted] for classes storing data instead of [Object].
```
Saying "unexpectedly" to me implies that it's a random occurrence, and can happen on its own. "To check for this" also doesn't connect to the previous sentence in the way that you want it to connect.

```suggestion
		[b]Note:[/b] Unlike references to a [RefCounted], references to an Object stored in a variable can become invalid without invalidating the variable itself. Because of this you should avoid checking such variables against [code]null[/code]. To reliably check for the Object's validity, use [method @GlobalScope.is_instance_valid]. It's recommended to inherit from [RefCounted] for classes storing data instead of [Object].
```

YuriSizov commented on doc/classes/Object.xml:
```diff
@@ -261,44 +323,44 @@
-				One can also pass additional parameters when emitting the signal with [method emit_signal]. The examples below show the relationship between those two types of parameters.
+				While all options have the same outcome ([code]button[/code]'s [signal BaseButton.button_down] signal will be connected to [code]_on_button_down[/code]), [b]option 3[/b] offers the best validation: it will print a compile-time error if either the [code]button_down[/code] [Signal] or the [code]_on_button_down[/code] [Callable] are not defined. On the other hand, [b]option 2[/b] only relies on  ...
+				[b]Binding and passing parameters:[/b]
+				The recommended syntax to bind parameters is by using [method Callable.bind], which returns a copy of the [Callable] with its parameters bound. For legacy reasons, the [method Signal.connect] method also allows passing the [code]binds[/code] as an [Array].
```
```suggestion
				The recommended syntax to bind parameters is by using [method Callable.bind], which returns a copy of the [Callable] with its parameters bound.
```

I don't think this is true anymore? Signal.connect only takes the callable and flags.

Mickeon replied:
Yeah, it's plainly wrong... That whole description under `connect` needs to be completely stripped away, but for now it's better just remove the outdated bits from there...

YuriSizov replied:
Sounds good!

YuriSizov commented on doc/classes/Object.xml:
```diff
@@ -514,89 +639,119 @@
 			<description>
-				Assigns a script to the object. Each object can have a single script assigned to it, which are used to extend its functionality.
-				If the object already had a script, the previous script instance will be freed and its variables and state will be lost. The new script's [method _init] method will be called.
+				Attaches [param script] to the object, and instantiate it. As a result, the script's [method _init] is called. A [Script] is used to extend the object's functionality.
```
```suggestion
				Attaches [param script] to the object, and instantiates it. As a result, the script's [method _init] is called. A [Script] is used to extend the object's functionality.
```

YuriSizov reviewed (changes requested):
I thought I'd approve it as is, but I think there are still a couple of changes that are worth to be made before merging. Not everything is severe, so I tried to note the ones where I just nitpick with comments. But there are also a couple of typos 🙂

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -38,78 +42,146 @@
-				Virtual method which can be overridden to customize the return value of [method get].
-				Returns the given property. Returns [code]null[/code] if the [param property] does not exist.
+				Override this method to customize the behavior of [method get]. Should return the given [param property]'s value, or [code]null[/code] if the [param property] does not exist.
+				Combined with [method _set] and [method _get_property_list], this method allows defining custom properties, which is particularly useful for editor plugins. Note that a property must be present in [method get_property_list], otherwise this method may not be called.
```
This was probably combined with the confusion/inconsistency I felt the other day regarding when `_get` was actually called.

Mickeon commented on doc/classes/Object.xml:
```diff
@@ -38,78 +42,146 @@
-				Called when the object is initialized in memory. Can be defined to take in parameters, that are passed in when constructing.
-				[b]Note:[/b] If [method _init] is defined with required parameters, then explicit construction is the only valid means of creating an Object of the class. If any other means (such as [method PackedScene.instantiate]) is used, then initialization will fail.
+				Called when the object's script is instantiated, oftentimes after the object is initialized in memory. Can be defined to take in parameters. This method is similar to a constructor in most programming languages.
+				[b]Note:[/b] If [method _init] is defined with [i]required[/i] parameters, explicit construction will be the only valid way to create an Object of the class. If any other means (such as [method PackedScene.instantiate]) are used, then initialization will fail.
```
Completely rewrote this instead of leeching off the previous description.

Mickeon commented:
Didn't go with the suggestions 1:1, but I'm particularly glad to have touched upon the `_init` one because it was explained very awkwardly for being such an important part of the engine.

YuriSizov reviewed (approved):
I'm still not a fan of "If any other means are used", but I won't die on that hill. Let's merge!

YuriSizov commented:
> "If any other... method?"

Well, I've given an example of how I'd rephrase it, but let's not. It's been built enough times already. I'm pressing the merge button!

Mickeon commented:
OKAY FINE MEGA MERGE IT IS.

YuriSizov commented:
Thanks a bunch!
